### PR TITLE
Normalize substs before resolving instance in `NoopMethodCall` lint

### LINKED ISSUE #102074: ICE(s?) with debug assertions enabled in rustc_trait_selection
### Meta
These are from the crater runs on https://github.com/rust-lang/rust/pull/101929

### Error output
There are quite a few crates in the crater run which hit this assertion, I'm just picking out 3 examples that look kind of representative to me. If you want more examples, search through the build-fail crates. The majority of failures are in there, for whatever reason.

https://crater-reports.s3.amazonaws.com/pr-101929/try%233bbb5242a8064de4ca3af4036745ddf603ddaa85/reg/EasyCrypto-0.8.1/log.txt
```
[INFO] [stderr] thread 'rustc' panicked at 'assertion failed: `(left == right)`
[INFO] [stderr]   left: `Binder(<[u8; _] as std::clone::Clone>, [])`,
[INFO] [stderr]  right: `Binder(<[u8; 16] as std::clone::Clone>, [])`', compiler/rustc_trait_selection/src/traits/codegen.rs:26:5
```
https://crater-reports.s3.amazonaws.com/pr-101929/try%233bbb5242a8064de4ca3af4036745ddf603ddaa85/reg/holochain_deterministic_integrity-0.0.13/log.txt
```
[INFO] [stderr] thread 'rustc' panicked at 'assertion failed: `(left == right)`
[INFO] [stderr]   left: `Binder(<to_local_types_test_nested::Nesting as holochain_deterministic_integrity::prelude::EnumVariantLen<#[hdk_to_coordinates(nested)]>>, [])`,
[INFO] [stderr]  right: `Binder(<to_local_types_test_nested::Nesting as holochain_deterministic_integrity::prelude::EnumVariantLen<0>>, [])`', compiler/rustc_trait_selection/src/traits/codegen.rs:26:5
```
https://crater-reports.s3.amazonaws.com/pr-101929/try%233bbb5242a8064de4ca3af4036745ddf603ddaa85/reg/pluser-3.0.0/log.txt
```
[INFO] [stderr] thread 'rustc' panicked at 'assertion failed: `(left == right)`
[INFO] [stderr]   left: `Binder(<accumulator::Accumulator as ink_lang::reflect::DispatchableConstructorInfo</// Holds a simple `i32` value that can be incremented and decremented.
[INFO] [stderr]     #[ink(storage)]
[INFO] [stderr]     pub struct Accumulator {
[INFO] [stderr]         value: i32,
[INFO] [stderr]     }>>, [])`,
[INFO] [stderr]  right: `Binder(<accumulator::Accumulator as ink_lang::reflect::DispatchableConstructorInfo<2611912030>>, [])`', compiler/rustc_trait_selection/src/traits/codegen.rs:26:5
```

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #102074

r? types

### CODE REVIEW

File: compiler/rustc_lint/src/noop_method_call.rs [hunk 1]
```diff
@@ -56,21 +56,22 @@ impl<'tcx> LateLintPass<'tcx> for NoopMethodCall {
...
-        let substs = cx.typeck_results().node_substs(expr.hir_id);
+        let substs = cx
+            .tcx
+            .normalize_erasing_regions(cx.param_env, cx.typeck_results().node_substs(expr.hir_id));
         if substs.needs_subst() {
             // We can't resolve on types that require monomorphization, so we don't handle them if
             // we need to perform substitution.
             return;
         }
-        let param_env = cx.tcx.param_env(trait_id);
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/noop_method_call.rs:71]: I don't think it made sense to use the param-env of the trait here...?

**Comment by matthiaskrgr** [comment]
This might also close https://github.com/rust-lang/rust/issues/97725 :3

File: compiler/rustc_lint/src/noop_method_call.rs [hunk 2]
```diff
@@ -56,21 +56,22 @@ impl<'tcx> LateLintPass<'tcx> for NoopMethodCall {
...
                 }
                 _ => return,
             },
             _ => return,
         };
-        let substs = cx.typeck_results().node_substs(expr.hir_id);
+        let substs = cx
+            .tcx
+            .normalize_erasing_regions(cx.param_env, cx.typeck_results().node_substs(expr.hir_id));
         if substs.needs_subst() {
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_lint/src/noop_method_call.rs:68]: `Instance::resolve` should not panic if `substs.needs_subst`, it should return a non `Ok(Some(_))` value. Removing that check should be ok.

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_lint/src/noop_method_call.rs]: You're right, and it also fixes a exactly the case that was missing a warning in the original noop-method-call ui test :smile_cat:

**Pull request merged**